explorer/templates.go

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2017, The dcrdata developers
+// See LICENSE for details.
+
+package explorer
+
+import (
+	"bytes"
+	"fmt"
+	"html/template"
+	"math"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/decred/dcrd/chaincfg"
+	humanize "github.com/dustin/go-humanize"
+)
+
+type pageTemplate struct {
+	file     string
+	template *template.Template
+}
+
+type templates struct {
+	templates map[string]pageTemplate
+	defaults  []string
+	folder    string
+	helpers   template.FuncMap
+}
+
+func newTemplates(folder string, defaults []string, helpers template.FuncMap) templates {
+	var defs []string
+	for _, file := range defaults {
+		defs = append(defs, filepath.Join(folder, file+".tmpl"))
+	}
+	return templates{
+		templates: make(map[string]pageTemplate),
+		defaults:  defs,
+		folder:    folder,
+		helpers:   helpers,
+	}
+}
+
+func (t templates) addTemplate(name string) error {
+	fileName := filepath.Join(t.folder, name+".tmpl")
+	files := append(t.defaults, fileName)
+	temp, err := template.New(name).Funcs(t.helpers).ParseFiles(files...)
+	if err == nil {
+		t.templates[name] = pageTemplate{
+			file:     fileName,
+			template: temp,
+		}
+	}
+	return err
+}
+
+func (t templates) reloadTemplates() error {
+	var errorStrings []string
+	for fileName := range t.templates {
+		err := t.addTemplate(fileName)
+		if err != nil {
+			errorStrings = append(errorStrings, err.Error())
+		}
+	}
+	if errorStrings == nil {
+		return nil
+	}
+	return fmt.Errorf(strings.Join(errorStrings, " | "))
+}
+
+func (t templates) execTemplateToString(name string, data interface{}) (string, error) {
+	temp, ok := t.templates[name]
+	if !ok {
+		return "", fmt.Errorf("Template %s not know", name)
+	}
+
+	var page bytes.Buffer
+	err := temp.template.ExecuteTemplate(&page, name, data)
+	return page.String(), err
+}
+
+var toInt64 = func(v interface{}) int64 {
+	switch vt := v.(type) {
+	case int64:
+		return vt
+	case int32:
+		return int64(vt)
+	case uint32:
+		return int64(vt)
+	case uint64:
+		return int64(vt)
+	case int:
+		return int64(vt)
+	case int16:
+		return int64(vt)
+	case uint16:
+		return int64(vt)
+	default:
+		return math.MinInt64
+	}
+}
+
+func makeTempHelpers(params *chaincfg.Params) template.FuncMap {
```

Comment:
just a nitpick, but I think it would be clearer if the function were named `makeTemplateFuncMap`.